### PR TITLE
Upgrade tests for KAFKA-13769

### PR DESCRIPTION
**Pull request by Gerrrr**
This PR adds upgrade tests for the changes in Foreign Key join protocol in KAFKA-13769. New upgrade tests ensure that FK joins keep working after a 2-bounce upgrade despite the protocol changes. I added the tests for all Kafka versions starting from 2.4 - the version that introduced FK joins.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by Gerrrr** [comment]
@mjsax Can you please review?

File: streams/upgrade-system-tests-31/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java [hunk 1]
```diff
@@ -63,21 +80,30 @@ public static void main(final String[] args) throws Exception {
         }));
     }
 
-    private static <KIn, VIn, KOut, VOut> ProcessorSupplier<KIn, VIn, KOut, VOut> printProcessorSupplier() {
+    private static void buildFKTable(final KStream<String, Integer> primaryTable,
+        final KTable<Integer, String> otherTable) {
```
Comment (mjsax) [inline streams/upgrade-system-tests-31/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]: nit: indention

File: tests/kafkatest/tests/streams/streams_upgrade_test.py [hunk 2]
```diff
@@ -37,6 +37,7 @@
 
 metadata_1_versions = [str(LATEST_0_10_0)]
 metadata_2_versions = [str(LATEST_0_10_1), str(LATEST_0_10_2), str(LATEST_0_11_0), str(LATEST_1_0), str(LATEST_1_1)]
+fk_join_versions = [str(V_2_4_0)]
```
Comment (mjsax) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py]: We only test with latest bug-fix releases, so we can use `LATEST_2_4` here.

Comment (mjsax) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: Should we test all versions 2.4 - 3.1 ?

**Review by mjsax — COMMENTED** [review]
Thanks for the PR. Can you trigger a system test run to ensure it's working before we merge it.

**Comment by Gerrrr** [comment]
@mjsax How can I trigger system tests in the CI?

File: tests/kafkatest/tests/streams/streams_upgrade_test.py (hunk 2 above)
Comment (Gerrrr) [reply tests/kafkatest/tests/streams/streams_upgrade_test.py]: Done in [d530a29](https://github.com/apache/kafka/pull/12122/commits/d530a29d1df6bb5302745c54ef08c83dc6c9c6de).

File: streams/upgrade-system-tests-31/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java (hunk 1 above)
Comment (Gerrrr) [reply streams/upgrade-system-tests-31/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]: Done in [d530a29](https://github.com/apache/kafka/pull/12122/commits/d530a29d1df6bb5302745c54ef08c83dc6c9c6de).

**Comment by Gerrrr** [comment]
Green CI run with system tests - https://jenkins.confluent.io/job/system-test-kafka-branch-builder/4897/.

**Pull request merged**